source/slang/slang-ir-specialize.cpp

```diff
@@ -211,7 +212,37 @@ struct SpecializationContext
         IRGeneric* genericVal,
         IRSpecialize* specializeInst)
     {
-        // First, we want to see if an existing specialization
+        // We need to fold the generic arguments here in order to uniquely identify
+        // which specializations need to be generated.
+        // The folding of the generic-arguments are deferred until the specialization
+        // step here, because the exact value of the generic-arguments can be unknown
+        // until the specialization. The exact values of the generic-arguments may come
+        // from the other modules, and they will be unknown until linking the modules.
+        //
+        UInt argCount = specializeInst->getArgCount();
+        {
+            IRBuilder builder(module);
+            for (UInt ii = 0; ii < argCount; ++ii)
+            {
+                IRUse* nextUse = nullptr;
+                IRInst* argInst = specializeInst->getArg(ii);
+                for (auto iuse = argInst->firstUse; iuse; iuse = nextUse)
```

Comment:
There is not need to iterate over the uses of argInst.

```
IRUse* argUse = specializeInst->getArgOperand(ii);
auto originalArg = argUse->get();
IRInst* foldedArg = tryConstantFoldInst(module, originalArg);
if (foldedArg == originalArg)
    continue;

specializeInst = builder.replaceOperand(argUse, foldedArg);
```

Reply:
I don't see a function called "getArgOperand()" anywhere.
Instead, I can do something like this if I don't need to iterate for uses.
```
            IRBuilder builder(module);
            for (UInt ii = 0; ii < argCount; ++ii)
            {
                IRInst* argInst = specializeInst->getArg(ii);
                IRUse* argUse = argInst->firstUse;
                auto originalArg = argUse->get();
                IRInst* foldedArg = tryConstantFoldInst(module, originalArg);
                if (foldedArg == originalArg)
                    continue;

                specializeInst = as<IRSpecialize>(builder.replaceOperand(argUse, foldedArg));
            }
```

Reply:
Just add it to that class:
```
IRUse* getArgOperand(Index i) { return getOperands() + 1 + i; }
```

Reply:
Note that argInst->firstUse is not necessarily the operand of the IRCall.

Reply:
I see.
That makes sense.